xml and jsonpath support (#12084)
home-assistant/home-assistant.github.io: source/_integrations/rest.markdown

  required: false
  type: [string, list]
json_attributes:
  description: A list of keys to extract values from a JSON dictionary result and then set as sensor attributes. If the endpoint returns XML with the "text/xml" content type, it will automatically be converted to JSON according to this [specification](https://www.xml.com/pub/a/2006/05/31/converting-between-xml-and-json.html)
  required: false
  type: [string, list]
json_path_attributes:

>   description: A [jsonpath](https://goessner.net/articles/JsonPath/) that refereances the location of the json_attributes in the JSON content.

Reviewer: ```suggestion
  description: A [JSONPath](https://goessner.net/articles/JsonPath/) that references the location of the `json_attributes` in the JSON content.
```

```
{% endraw %}

> [JSONPlaceholder](https://jsonplaceholder.typicode.com/) provides sample JSON data for testing. In the below example, jsonpath locates the attributes in the JSON document. [JSONPath Online Evaluator](https://jsonpath.com/) provides a tool to test your jsonpath. If the endpoint returns XML, it will be converted to JSON using xmltodict before searching for attributes.

Reviewer: ```suggestion
[JSONPlaceholder](https://jsonplaceholder.typicode.com/) provides sample JSON data for testing. In the below example, JSONPath locates the attributes in the JSON document. [JSONPath Online Evaluator](https://jsonpath.com/) provides a tool to test your JSONPath. If the endpoint returns XML, it will be converted to JSON using `xmltodict` before searching for attributes.
```

> {% raw %}

Reviewer: ```suggestion
{% raw %}

```

```yaml
sensor:
  - platform: rest
    name: JSON users
    json_attributes_path: "$.[0].address"
    json_attributes:
      - street
      - suite
      - city
      - zipcode
    resource: https://jsonplaceholder.typicode.com/users

>     value_template: '{{ value_json.name }}'

Reviewer 2: Should be `{{ value_json[0].name }}`

Author: Thanks @arsaboo.  Will adjust this weekend as I'm traveling today.

```

> {% endraw %}

Reviewer: ```suggestion

{% endraw %}
```

        unit_of_measurement: '°C'
```
{% endraw %}

> The below example allows shows how to extract multiple values from a dictionary with `json_attributes` and `json_attributes_path` from the XML of a Steamist Steambath WiFi interface and use them to create a switch and multiple sensors without having to poll the endpoint numerous times.

Reviewer: ```suggestion
The below example allows shows how to extract multiple values from a dictionary with `json_attributes` and `json_attributes_path` from the XML of a Steamist Steambath Wi-Fi interface and use them to create a switch and multiple sensors without having to poll the endpoint numerous times.
```

> In the below example `json_attributes_path` is set to "$.response" which is the location of the the "usr0", "pot0", ... attributes used for `json_attributes`.

Reviewer: ```suggestion
In the below example `json_attributes_path` is set to `$.response` which is the location of the `usr0`, `pot0`, ... attributes used for `json_attributes`.
```

> {% raw %}

Reviewer: ```suggestion
{% raw %}

```

        friendly_name: "Steam Time Remaining"
        value_template: '{{ states.sensor.steam_system_data.attributes["time0"] }}'
        unit_of_measurement: "minutes"

switch: 
  - platform: template
    switches:
      steam:
        value_template: '{{ states.sensor.steam_system_data.attributes["usr0"] | int >= 1 }}'
        turn_on:
          - service: rest_command.set_steam_led
            data:
               led: 6
          - service: homeassistant.update_entity
            data:
               entity_id: sensor.steam_system_data
          - delay: 00:00:15
          - service: homeassistant.update_entity
            data:
               entity_id: sensor.steam_system_data
        turn_off:
          - service: rest_command.set_steam_led
            data:
               led: 7
          - service: homeassistant.update_entity
            data:
               entity_id: sensor.steam_system_data
          - delay: 00:00:15
          - service: homeassistant.update_entity
            data:
               entity_id: sensor.steam_system_data
        friendly_name: Steam 

rest_command:  
  set_steam_led:
    url: http://192.168.1.105/leds.cgi?led={{ led }}

> 

Reviewer: ```suggestion
```

```

> {% endraw %}

Reviewer: ```suggestion

{% endraw %}
```

For reference, the XML content of endpoint shown above example is below:

> {% raw %}

Reviewer: ```suggestion
```

```xml
<?xml version="1.0" encoding="utf-8"?>
 
 <response>
 	<scan>0</scan>
	<ver>12556</ver>
	<count>48</count>
	<ssid>alexander</ssid>
	
	<bss>
		<valid>0</valid>
		<name>0</name>
		<privacy>0</privacy>
		<wlan>0</wlan>
		<strength>0</strength>
	</bss>
	
	<led0>0</led0>
	<led1>0</led1>
	<led2>0</led2>
	<led3>0</led3>
	<led4>0</led4>
	<led5>0</led5>
	<led6>0</led6>
	<led7>0</led7>
	<btn0>up</btn0>
	<btn1>up</btn1>
	<btn2>up</btn2>
	<btn3>up</btn3>
	<pot0>0</pot0>
	<usr0>0</usr0>
	<temp0>0x73XF0x73XF</temp0>
	<time0> 0</time0>

> 

Reviewer: we don't need that many new lines here. Let's keep it tidy and make it a single new-line.

Author: fixed

	
 </response>
```

> {% endraw %}

Reviewer: ```suggestion
```
